howso/boundary_values.amlg

```diff
@@ -0,0 +1,814 @@
+(null
+
+	;Compute the values for context features that would perturb the prediction action values
+	#!ComputeBoundaryValues
+	(let
+		(assoc
+			context_feature_index_map (zip context_features (indices context_features))
+			action_feature_index_map (zip action_features (indices action_features))
+
+			boundary_action_condition (get details "boundary_value_action_outcome")
+			boundary_value_features (get details "boundary_value_context_features")
+		)
+
+		(declare (assoc
+			feature_deviations  (get hyperparam_map "featureDeviations")
+			feature_weights (get hyperparam_map "featureWeights")
+		))
+
+		(if (get hyperparam_map "subtraineeName")
+			(call !UseDynamicDeviationsAndWeights (assoc
+				context_features context_features
+				context_values context_values
+				hyperparam_map hyperparam_map
+			))
+		)
+
+		;define eval_function, a piece of code that can be called which will either return:
+		; 1. a boolean indicating if `new_action_values` represent new action values that either
+		;  crossed the boundary or satisfied the user given conditions
+		; 2. a value representing the ratio of the difference in context & action values
+		;   (`new_context_values` & `new_action_values`) to the difference in just the context values
+		(declare (assoc
+			eval_function
+				(if boundary_action_condition
+					;function that evaluates new_action_values to determine if they meet the user-specified conditions
+					(lambda
+						(apply "and" (values
+							(map
+								(lambda
+									(let
+										(assoc
+											feat (current_index 1)
+											condition (current_value 1)
+											feat_value (get new_action_values (get action_feature_index_map (current_index 1)))
+										)
+
+										(if (contains_index !nominalsMap feat)
+											;nominals block, users can specify one value that it must match or a list of possible values
+											(if (~ (list) condition)
+												(contains_value condition feat_value)
+
+												(= condition feat_value)
+											)
+
+											;continuous, users specify a tuple of the accepted range
+											(and
+												(>= feat_value (first condition) )
+												(<= feat_value (last condition) )
+											)
+										)
+									)
+
+								)
+								boundary_action_condition
+							)
+						))
+					)
+
+					;no specified boundary, define a function that will indicate a boundary based on action values
+					(let
+						(assoc
+							num_nominal_features (size (filter (lambda (= "nominal" (get !featureAttributes [(current_value 1) "type"]))) action_features))
```

Comment:
(contains !NominalsMap (current_value)) is simpler and faster than checking 2 layers deep into attributes and comparing to a string.

Reply:
will do